akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ReceiveBuilder.scala

```diff
@@ -196,24 +212,61 @@ object ReceiveBuilder {
  */
 @InternalApi
 private final class BuiltReceive[T](
-    messageHandlers: List[ReceiveBuilder.Case[T, T]],
-    signalHandlers: List[ReceiveBuilder.Case[T, Signal]])
+    messageHandlers: Array[ReceiveBuilder.Case[T, T]],
+    signalHandlers: Array[ReceiveBuilder.Case[T, Signal]])
     extends Receive[T] {
   import ReceiveBuilder.Case
 
-  override def receiveMessage(msg: T): Behavior[T] = receive[T](msg, messageHandlers)
-
-  override def receiveSignal(msg: Signal): Behavior[T] = receive[Signal](msg, signalHandlers)
+  override def receiveMessage(msg: T): Behavior[T] = receive[T](msg, messageHandlers, 0)
 
+  override def receiveSignal(msg: Signal): Behavior[T] = receive[Signal](msg, signalHandlers, 0)
   @tailrec
-  private def receive[M](msg: M, handlers: List[Case[T, M]]): Behavior[T] =
-    handlers match {
-      case Case(cls, predicate, handler) :: tail =>
-        if ((cls.isEmpty || cls.get.isAssignableFrom(msg.getClass)) && (predicate.isEmpty || predicate.get.test(msg)))
-          handler(msg)
-        else receive[M](msg, tail)
-      case _ =>
-        Behaviors.unhandled
+  protected final def receive[M](msg: M, handlers: Array[Case[T, M]], idx: Int): Behavior[T] = {
+    if (handlers.length == 0) {
+      Behaviors.unhandled[T]
+    } else {
+      val Case(cls, predicate, handler) = handlers(idx)
+      if ((cls.isEmpty || cls.get.isAssignableFrom(msg.getClass)) && (predicate.isEmpty || predicate.get.test(msg)))
+        handler(msg)
+      else if (idx == handlers.length - 1)
+        Behaviors.unhandled[T]
+      else
+        receive(msg, handlers, idx + 1)
     }
+  }
+}
 
+/**
+ * Receive type for [[AbstractBehavior]]
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private final class BuiltClassifiedReceive[T](
+    classifiedMessageHandlers: Map[Class[_ <: T], Array[ReceiveBuilder.Case[T, T]]],
+    signalHandlers: Array[ReceiveBuilder.Case[T, Signal]])
+    extends Receive[T] {
+  import ReceiveBuilder.Case
+
+  override def receiveMessage(msg: T): Behavior[T] = classifiedMessageHandlers.get(msg.getClass) match {
+    case Some(handlers) => receive(msg, handlers, 0)
+    case None           => Behaviors.unhandled[T]
+  }
+
+  override def receiveSignal(sig: Signal): Behavior[T] = receive(sig, signalHandlers, 0)
+
+  @tailrec
+  private def receive[M](msg: M, handlers: Array[Case[T, M]], idx: Int): Behavior[T] = {
+    if (handlers.length == 0) {
+      Behaviors.empty[T]
+    } else {
+      val Case(cls, predicate, handler) = handlers(idx)
+      if ((cls.isEmpty || cls.get.isAssignableFrom(msg.getClass)) && (predicate.isEmpty || predicate.get.test(msg)))
+        handler(msg)
+      else if (idx == handlers.length - 1)
+        Behaviors.unhandled[T]
+      else
+        receive(msg, handlers, idx + 1)
+    }
+  }
```

Comment:
One thing that is very important with these two new optimized receive logics is to benchmark and make sure that there really is a performance benefit (and if we can verify that, maybe also at what number of handlers it starts being beneficial). 

The `akka-bench-jmh` module is where we have our jmh benchmarks, I think there is no existing bench we can use for this but that it this will require a new benchmark, testing invoking the built receive directly without running it inside an actor. You'll likely have to do some exploratory benchmarking where you run a few different sets of receive cases with both the array based one and this optimized case based one to see how they differ in throughput.

Reply:
I will try that in this weekend.